DiscordBot/commands/tournaments/mappool/publish.ts

```diff
@@ -0,0 +1,143 @@
+    const user = await User.findOne({
+        where: {
+            discord: {
+                userID: m instanceof Message ? m.author.id : m.user.id,
+            }
+        }
+    })
+    if (!user) {
+        await loginResponse(m);
+        return;
+    }
+
+    const poolRegex = /-p (\S+)/;
+    const poolText = m instanceof Message ? m.content.match(poolRegex) ?? m.content.split(" ")[1] : m.options.getString("pool");
+    if (!poolText) {
+        if (m instanceof Message) m.reply("Missing parameters. Please use `-p <pool>` or `<pool>`. If you do not use the `-` prefixes, the order of the parameters is important.");
+        else m.editReply("Missing parameters. Please use `-p <pool>` or `<pool>`. If you do not use the `-` prefixes, the order of the parameters is important.");
+        return;
+    }
+
+    const pool = typeof poolText === "string" ? poolText : poolText[0];
+
+    const mappool = await fetchMappool(m, tournament, pool, true, true, true);
+    if (!mappool) 
+        return;
+
+    // Confirmations
+    if (!mappool.isPublic && mappool.stage.timespan.start.getTime() - Date.now() > 1000 * 60 * 60 * 24 * 7) {
+        const confirm = await confirmCommand(m, "This mappool is more than a week away from the stage's start date. Are you sure you want to publish it?");
+        if (!confirm)
+            return;
+    } 
+    if (!mappool.isPublic && mappool.slots.some(slot => slot.maps.some(map => !map.beatmap && !map.customBeatmap))) {
+        const confirm = await confirmCommand(m, "This mappool still contains empty slots. Are you sure you want to publish it?");
+        if (!confirm)
+            return;
+    }
+    if (mappool.isPublic && mappool.stage.timespan.start.getTime() < Date.now()) {
+        const confirm = await confirmCommand(m, "This mappool's stage has already started. Are you sure you want to privatize it?");
+        if (!confirm)
+            return;
+    }
+
+    if (mappool.isPublic) {
+        // Reset link before making it private
+        await buckets.mappacks.deleteObject(mappool.s3Key!);
+        mappool.s3Key = mappool.link = mappool.linkExpiry = null;
+        mappool.isPublic = false;
+
+        await mappool.save();
+
+        if (m instanceof Message) m.reply(`**${mappool.name.toUpperCase()}** (${mappool.abbreviation.toUpperCase()}) is now private`);
+        else m.editReply(`**${mappool.name.toUpperCase()}** (${mappool.abbreviation.toUpperCase()}) is now private`);
+
+        await mappoolLog(tournament, "publish", user, `**${mappool.name.toUpperCase()}** (${mappool.abbreviation.toUpperCase()}) is now private`);
+
+        return;
+    }
+
+    const s3Key = mappool.s3Key || `${randomUUID()}/${tournament.abbreviation.toUpperCase()}${tournament.year} ${mappool.abbreviation.toUpperCase()}.zip`;
+
+    const mappack = await buckets.mappacksTemp.getObject(s3Key);
+    if (mappack) {
+        await buckets.mappacksTemp.deleteObject(s3Key);
+        await buckets.mappacks.putObject(s3Key, mappack);
```

Comment:
i implemented `copyObject` for exactly this use case!

Reply:
O